### ISSUE TITLE
Notification script

### ISSUE THREAD

**Issue by domibarton** [sabnzbd/sabnzbd#435]
Hi there!

First of all, awesome job! :) I love SABnzbd!

Have you ever thought about a "custom" notification script? I know there are custom post processing scripts per category and I already use them for SickBeard. But it would be great to get a custom notification script. If I'm not completely wrong I can use Email, Growl, Prowl, Pushover and Pushbullet. Why not adding a new option to execute a custom script? 

I want to use it to notify a Slack channel. Of course it would be even more awesome to get a specific Slack notification option.

What do you think? If you like the idea I'll have a look if I can integrate such an option and send you a PR.

Cheers
Domi

**Comment by shypike** [comment]
A custom notification script would probably be better than a specific Slack option.
We don't want SABnzbd to become an exhibition of notification services.
Most of the real code is in sabnzbd/growler.py and there is the UI in sabnzbd/interface.py and the Config->Notifications template.

**Comment by domibarton** [comment]
OK I understand. Makes sense :)
I'll have a look into it and try to build something.

**Comment by Safihre** [comment]
Also take a look at Sonarr and Couchpotato for TV shows and movies, they have a really big amount of notification services build in. 
Of course only works for TV shows and movies then :)

**Comment by domibarton** [comment]
There it is. Was quite a journey to be honest :) 

I think we could probably clean up a lot of the code because there's repetition of code everywhere. I tried to clean it up, but that requires more than just a day ^^ Would be great to get a proper abstract `Notification` class to inherit all concrete notification implementations from. 

The "management" of the priorities and additional custom parameters per notification type will become much easier. Something like that:

```python
class ProwlNotification(Prowl):
    ''' Notification class for Prowl. '''
    help_text = 'Requires a Prowl account'
    parameters = (
        NotificationParameter('apikey', 'API key for prowl', 'Personal API key for Prowl', required=True),
    )

    def notify(self, parameters):
        ''' Sends Prowl notification. '''

        # do some business logic here
```

The Notification class can handle then all the use cases, for example verifications and sending test messages (i.e. getting parameters from GET/POST instead from `sabnzbd.cfg`).

In this case someone could easily add new Notification types by just creating a new `Notification`-based class. Of course the HTML on the config front-end has to be rendered dynamically, but that's not very hard.

What do you think?

**Comment by domibarton** [comment]
@Safihre yeah I already use SickBeard and CouchPotato, but I want to send notifications into Slack. CouchPotato supports Slack, but SickBeard doesn't. So I thought I can build it into SickBeard or SABnzbd, but I like the idea of configuring notifications only on one single point for all kind of downloads... so SABnzbd it is :)

**Comment by shypike** [comment]
Closed by commit ccba56e073049c3c092f2be179b0a56d0ce4e798

**Comment by seanvree** [comment]
hey @domibarton  did you ever figure out the SLACK notifications?  I know  you can do an email webhook and do it that way...just wondering. 

THX!

**Comment by Safihre** [comment]
You should use nzbnotify Notification Script, it has the support built-in: https://github.com/caronc/nzb-notify

We also link this from several wiki pages :)

**Comment by domibarton** [comment]
@seanvree I did, but I'm using a self-built Python script… I think @caronc's solution as mentioned by @Safihre might be a better fit for you. 

I'm also using [Plex](https://www.plex.tv/) as my media server and I found out that @JonnyWong16's [PlexPy](https://github.com/JonnyWong16/plexpy) supports Slack notifications out of the box. So if you're using Plex, I'd recommend you have a look at PlexPy - it's really nice and it will send notifications as soon as Plex indexed something new. There are also a lot of stats included in PlexPy - so a win-win ;)

If you don't use Plex, I'd recommend to have a look at `nzb-notify` :)
